### PR TITLE
[nnc] Infer device type from nodes if inputs are all scalars

### PR DESCRIPTION
**Pull request by bertmaher**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59430 [nnc] Infer device type from nodes if inputs are all scalars**

With constant support added, we can now have fusion groups with only
scalar inputs.  So, we need to get the device type from the nodes in the graph
rather than just the inputs.

Differential Revision: [D28891989](https://our.internmc.facebook.com/intern/diff/D28891989/)

### CODE REVIEW

**Review by navahgar — APPROVED** [review]
LGTM

File: test/test_jit_fuser_te.py
Comment (ZolotukhinM) [inline test/test_jit_fuser_te.py:1845]: Dynamic shapes are not printed as "Dynamic", they are printed as "Tensor".
In general, I'm a bit suspicious about this test - do you mind elaborating what we're trying to check here? Do we really need two functions to check how scalars are handled? Can we simply parse a graph with scalar inputs from text? Another reason why I'm suspicious is that AFAICT we shouldn't be fusing scalars at all until #59279 lands.

File: torch/csrc/jit/tensorexpr/kernel.cpp
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/kernel.cpp:124]: What if a graph has both cpu and cuda nodes? Should we error out?

File: test/test_jit_fuser_te.py
Comment (bertmaher) [reply test/test_jit_fuser_te.py:1845]: I lifted this test verbatim from `test_tracer.py test_trace_of_script`, so that we'd get OSS coverage of the fuser in that case.  But I supposed I can write a more minimal test case here.

I guess I usually prefer writing python code for tests rather than using the IR parser, but I can see the advantages.

I don't think we've ever prevented scalar fusion have we?  What's new here is that the *only* input to the graph is a scalar, because the tensor operations are performed on a constant.

**Comment by bertmaher** [comment]
@ZolotukhinM I think these changes should address your feedback

**Comment by ZolotukhinM** [comment]
I think it looks better and I'm ok with committing it as-is (esp. to unblock the next PR), but I still think this test is not great. It's hard for me to tell what's going on: how does the input graph look like, what's getting fused, where the constant and the scalars are coming from, why do we have `if` there, etc.

Note that if you simply need to inject a constant tensor into IR, you could insert a `torch.ones` (like you already do) and then run a constant-prop pass on the graph, like it's done here for instance:
https://github.com/pytorch/pytorch/blob/2119efd234ff80b2a86b5cd9d15fb29d1e18de49/test/cpp/tensorexpr/test_kernel.cpp#L1220-L1233
(this is a C++ test, but the same can be achieved in python obviously - and we don't have to parse a graph from text for this to work either: scripting or tracing would work too)

File: torch/csrc/jit/tensorexpr/kernel.cpp
Comment (ZolotukhinM) [reply torch/csrc/jit/tensorexpr/kernel.cpp:125]: Yeah, fuser would probably take care of that, but I was thinking about the AOT scenario where we don't have the fuser pass safeguard :)

File: test/test_jit_fuser_te.py
Comment (ZolotukhinM) [reply test/test_jit_fuser_te.py]: > I don't think we've ever prevented scalar fusion have we?

We did:
https://github.com/pytorch/pytorch/blob/2119efd234ff80b2a86b5cd9d15fb29d1e18de49/torch/csrc/jit/passes/tensorexpr_fuser.cpp#L850-L852

**Comment by bertmaher** [comment]
So I think the reason this is confusing is because it's not obvious how the test function gets sliced up into a fusion group and why that fusion group is problematic.  But I just realized that I can write a simpler test that gets at the exact graph; I had just had trouble getting it to properly torchscript when I tried earlier.  Update incoming.

**Comment by bertmaher** [comment]
Voilà:
```
def eager(b: float):
    a = torch.ones(1)
    return a * b
```

The problem is that (1) we form a fusion group consisting of the Constant(1) and aten::mul, but (2) the only input of that fusion group is a float, which has no device type.

**Comment by bertmaher** [comment]
Thank you for helping me to simplify this :-)

**Pull request closed without merging**
